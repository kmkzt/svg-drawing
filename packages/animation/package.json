{
  "name": "@svg-drawing/animation",
  "version": "4.0.0-beta.5",
  "license": "MIT",
  "description": "svg drawing library.",
  "author": "kmkzt<info.pscreator@gmail.com>",
  "repository": "github:kmkzt/svg-drawing",
  "homepage": "https://github.com/kmkzt/svg-drawing/tree/master/packages/animation#readme",
  "bugs": {
    "url": "https://github.com/kmkzt/svg-drawing/issues",
    "email": "info.pscreator@gmail.com"
  },
  "keywords": [
    "svg",
    "drawing",
    "animation"
  ],
  "types": "lib/index.d.ts",
  "main": "lib/index.cjs.js",
  "module": "lib/index.esm.js",
  "browser": "lib/index.umd.js",
  "files": [
    "lib/**/*"
  ],
  "scripts": {
    "dev": "NODE_ENV=development rollup -w -c",
    "dev:tsc": "NODE_ENV=development tsc -w",
    "build": "npm-run-all lib:*",
    "lib:clear": "rimraf lib/*",
    "lib:rollup": "NODE_ENV=production rollup -c",
    "lib:tsc": "NODE_ENV=production tsc",
    "typecheck": "tsc --noEmit",
    "lint": "eslint ./src --ext .js,.ts,.tsx;",
    "fmt": "yarn lint --fix",
    "prepare": "yarn build"
  },
  "optionalDependencies": {
    "@svg-drawing/core": "4.0.0-beta.5"
  },
  "publishConfig": {
    "access": "public"
  },
  "gitHead": "6811d0236e51e0204ffbf64a6124ca5469079373"
}
