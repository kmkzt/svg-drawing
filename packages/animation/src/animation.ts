import {
  Renderer,
  RendererOption,
  Path,
  camel2kebab,
  download,
  roundUp,
  svg2base64,
  createSvgElement,
  createSvgChildElement,
} from '@svg-drawing/core'

export interface AnimationOption extends RendererOption {
  ms: number
}
export type FrameAnimation = (origin: Path[], loopIndex?: number) => Path[]

export class SvgAnimation extends Renderer {
  public ms: number
  private _stopId: number
  private _stop: (() => void) | null
  private _anim: FrameAnimation | null
  private _restorePaths: Path[]
  private _framesNumber: number | undefined

  /**d
   * Releation animate element
   * TODO: add easing option
   */
  private _repeatCount: string
  constructor(
    el: HTMLElement,
    { background, ms }: AnimationOption = { ms: 60 }
  ) {
    super(el, { background })
    this.ms = ms
    this._stop = null
    this._anim = null
    this._restorePaths = []
    this._stopId = 0
    this._repeatCount = 'indefinite'
  }

  /**
   * @param {FramaAnimation} fn
   * @param {{ frame?: number; repeat?: number }} opts
   * `frame` is the number of frames to animate
   * `repeat` is related for rapeatCount of animate element attribute.
   */
  public setAnimation(
    fn: FrameAnimation,
    {
      frames,
      repeatCount,
      ms,
    }: { frames?: number; repeatCount?: number | string; ms?: number } = {}
  ): void {
    this._anim = fn
    this._framesNumber = frames
    this._repeatCount = repeatCount ? `${repeatCount}` : 'indefinite'
    if (ms) this.ms = ms
  }

  public stop(): boolean {
    if (this._stop) {
      this._stop()
      return true
    }
    return false
  }

  public restore(): void {
    this.paths = this._restorePaths
    this.update()
  }

  public generateFrame(index?: number): Path[] {
    if (!this._anim) return this.paths
    return this._anim(
      this._restorePaths.map((p) => p.clone()),
      index
    )
  }

  public start(): void {
    // If do not this first, this cannot get the number of frames well.
    this.stop()
    this._registerRestorePaths()

    let index = 0
    let start: number | undefined
    const ms = this.ms
    const loopCount: number = this._getFramesNumber()
    const frame: FrameRequestCallback = (timestamp) => {
      if (ms !== this.ms) {
        this.restore()
        this.start()
        return
      }
      if (!start || timestamp - start > ms) {
        start = timestamp
        this.paths = this.generateFrame(index)
        this.update()
        index = index > loopCount ? 0 : index + 1
      }
      this._stopId = requestAnimationFrame(frame)
    }
    this._stopId = requestAnimationFrame(frame)
    this._stop = () => {
      cancelAnimationFrame(this._stopId)
      this._stop = null
    }
  }

  public toAnimationElement(): SVGSVGElement {
    // If the animation is stopped, read the currently displayed Svg data.
    // If stopped in the middle, SVG in that state is displayed
    if (!this._stop) {
      this._registerRestorePaths()
    }

    const loopNumber = this._getFramesNumber()
    const animPathsList: Path[][] = Array(loopNumber)
      .fill(null)
      .map((_: any, i: number) => this.generateFrame(i))

    const dur = loopNumber * (this.ms > 0 ? this.ms : 1) + 'ms'
    const t = 1 / loopNumber
    const keyTimes = `0;${Array(loopNumber)
      .fill(undefined)
      .map((_, i) => roundUp((i + 1) * t, 4) + '')
      .join(';')}`

    const createAnimationElement = (
      origin: Path,
      attributeName: string,
      defaultValue: string,
      getValue: ({
        origin,
        path,
      }: {
        origin: Path
        path?: Path
      }) => string | undefined
    ): SVGElement | null => {
      const animValues = animPathsList.map((ap) => {
        const path = ap.find((p) => p.attrs.id === origin.attrs.id)
        return getValue({ origin, path }) || defaultValue
      })
      // return null if value is same all.
      if (animValues.every((v) => v === defaultValue)) return null

      return createSvgChildElement('animate', {
        dur,
        keyTimes,
        attributeName,
        repeatCount: this._repeatCount,
        values: [defaultValue, ...animValues].join(';'),
      })
    }
    const animEls = this._restorePaths.map((p) => {
      const pEl = p.toElement()
      const dAnimEl = createAnimationElement(
        p,
        'd',
        p.getCommandString(),
        ({ origin, path }) =>
          path && path.commands.length > 0
            ? path.getCommandString()
            : origin.commands[0].toString()
      )
      if (dAnimEl) pEl.appendChild(dAnimEl)

      // TODO: Check attribute key and value.
      const { id, ...attrs } = p.attrs // exclude id
      Object.entries(attrs).map(
        ([propertyName, val]: [string, string | undefined]) => {
          if (!val) return
          const attrName = camel2kebab(propertyName)
          const aEl = createAnimationElement(
            p,
            attrName,
            val,
            ({ path }) => path?.attrs[propertyName]
          )
          if (aEl) pEl.appendChild(aEl)
        }
      )

      return pEl
    })

    const size = { width: String(this.width), height: String(this.height) }
    const bgEl = this.background
      ? [createSvgChildElement('rect', { ...size, fill: this.background })]
      : []
    return createSvgElement(
      { width: String(this.width), height: String(this.height) },
      bgEl.concat(animEls)
    )
  }

  /**
   * @param filename
   * TODO: Support gif and apng
   */
  public downloadAnimation(filename?: string): void {
    download({
      data: svg2base64(this.toAnimationElement().outerHTML),
      extension: 'svg',
      filename,
    })
  }

  /**
   * @return {number}
   * Default value is total of commands length.
   */
  private _getFramesNumber(): number {
    return this._framesNumber && this._framesNumber > 0
      ? this._framesNumber
      : this._restorePaths.reduce((l, p) => l + p.commands.length, 0)
  }

  private _registerRestorePaths() {
    this._restorePaths = this.clonePaths().map((p, i) => {
      p.attrs.id = `t${i}`
      return p
    })
  }
}
