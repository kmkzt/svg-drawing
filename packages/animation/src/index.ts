import { Point } from '@svg-drawing/core'
import type {
  PathClass,
  PathAttributes,
  FrameAnimation,
} from '@svg-drawing/core'

export class ShakeFrame implements FrameAnimation {
  public readonly loops = 10
  constructor(private range: number = 5) {
    this.animation = this.animation.bind(this)
  }

  private randomNumber() {
    return Math.random() * this.range - this.range / 2
  }

  public animation(paths: PathClass[]) {
    for (let i = 0; i < paths.length; i += 1) {
      paths[i].setCommands(
        paths[i].absoluteCommands.map((c) => {
          const update = c.clone()

          update.points = c.points.map((po) =>
            po.add(new Point(this.randomNumber(), this.randomNumber()))
          )
          return update
        })
      )
    }
    return paths
  }
}

export class AttributeFrame implements FrameAnimation {
  constructor(private attributesList: PathAttributes[]) {
    this.animation = this.animation.bind(this)
  }

  get loops(): number {
    return this.attributesList.length
  }

  animation(paths: PathClass[], key: number) {
    const attr = this.attributesList[key]
    if (!attr) return paths

    for (let i = 0; i < paths.length; i += 1) {
      paths[i].updateAttributes(attr)
    }
    return paths
  }
}

export class DrawFrame implements FrameAnimation {
  constructor(private paths: ReadonlyArray<PathClass>) {
    this.animation = this.animation.bind(this)
  }

  get loops(): number {
    return this.paths.reduce((acc, p) => acc + p.absoluteCommands.length, 0)
  }

  animation(paths: PathClass[], key: number) {
    let count = key
    const update = []
    for (let i = 0; i < paths.length; i += 1) {
      const path = paths[i]
      const vertexLength = path.absoluteCommands.length

      path.setCommands(path.absoluteCommands.slice(0, count))
      update.push(path)

      count -= vertexLength
      if (count < 0) break
    }
    return update
  }
}
