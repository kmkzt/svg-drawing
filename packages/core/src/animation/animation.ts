import { roundUp, camel2kebab } from '../utils'
import type {
  PathClass,
  AnimationOption,
  FrameAnimation,
  AnimateAttribute,
  AnimateObject,
  ElementClass,
} from '../types'

const createAnimateAttributeValues = (
  animationPaths: (PathClass | undefined)[],
  attributeName: string,
  defaultValue: string
): string[] =>
  animationPaths.map((p) => p?.attrs[attributeName] || defaultValue)

const createAnimateCommandValues = (
  animationPaths: (PathClass | undefined)[],
  originPath: PathClass
): string[] =>
  animationPaths.map(
    (p) => p?.getCommandString() || originPath.relativeCommands[0].toString()
  )

export class Animation {
  public ms: number
  public originPaths: PathClass[]
  public generator?: Generator<PathClass[], void, unknown>
  private _frame: FrameAnimation | null = null
  private _repeatCount?: number
  constructor({ ms }: AnimationOption = { ms: 60 }) {
    this.ms = ms
    this.originPaths = []
  }

  /**
   * `frame` is the number of frames to animate `repeat` is related for
   * repeatCount of animate element attribute.
   *
   * @param {FramaAnimation} fn
   * @param {{ frame?: number; repeat?: number }} opts
   */
  public setup(
    frame: FrameAnimation,
    { repeatCount, ms }: { repeatCount?: number; ms?: number } = {}
  ): void {
    this._frame = frame

    this._repeatCount = repeatCount || undefined
    if (ms) this.ms = ms

    this.generator = this.setupGenerator()
  }

  public getFrame(key: number): PathClass[] {
    const paths = this.originPaths.map((p) => p.clone())

    if (!paths.length || !this._frame) return paths

    return this._frame.animation(paths, key)
  }

  public restorePaths() {
    return this.originPaths.map((p) => p.clone())
  }

  private *setupGenerator() {
    let loopId = 0
    let repeatCount = 0
    if (!this._frame) return

    while (!this._repeatCount || repeatCount < this._repeatCount) {
      if (loopId > this._frame.loops) {
        repeatCount += 1
        loopId = 0
      } else {
        loopId += 1
      }

      yield this.getFrame(loopId)
    }
  }

  public initialize(paths: ReadonlyArray<ElementClass>) {
    this.originPaths = paths.map((p) => p.clone())
    this.generator = this.setupGenerator()
    return this
  }

  public toJson(): AnimateObject[] {
    if (!this._frame) return []

    const loop = this._frame.loops
    const frameLoop = Array(this._frame.loops).fill(null)

    const animPathsList: PathClass[][] = frameLoop.map((_: any, i: number) =>
      this.getFrame(i)
    )

    const baseAttrs = {
      repeatCount: `${this._repeatCount || `indefinite`}`,
      dur: loop * (this.ms > 0 ? this.ms : 1) + 'ms',
      keyTimes: frameLoop.reduce(
        (acc: string, _, i) => acc + ';' + roundUp((i + 1) * (1 / loop), 4),
        '0'
      ),
    }

    return this.originPaths.map((basePath) => {
      const { key, attributes } = basePath.toJson()

      const animPaths = animPathsList.map((ap) =>
        ap.find((p) => p.key === basePath.key)
      )

      // TODO: Check attribute key and value.
      const animates: AnimateAttribute[] = Object.keys(attributes)
        .sort()
        .reduce((animateAttrs: AnimateAttribute[], attributeName: string) => {
          const defaultValue = attributes[attributeName]
          if (!defaultValue) return animateAttrs

          const values =
            attributeName === 'd'
              ? createAnimateCommandValues(animPaths, basePath)
              : createAnimateAttributeValues(
                  animPaths,
                  attributeName,
                  defaultValue
                )
          if (values.every((v) => v === defaultValue)) return animateAttrs

          return [
            ...animateAttrs,
            {
              ...baseAttrs,
              attributeName: camel2kebab(attributeName),
              values: [...values, defaultValue].join(';'),
            },
          ]
        }, [])

      return {
        key,
        animates,
      }
    })
  }
}
