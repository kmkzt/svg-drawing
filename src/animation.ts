import { Renderer, RendererOption } from './renderer'
import { Path } from './svg'
import { camel2kebab } from './utils/camel2kebab'
import { download } from './utils/download'
import { roundUp } from './utils/roundUp'
import { svg2base64 } from './utils/svg2base64'
import {
  createSvgElement,
  createSvgChildElement,
} from './utils/createSvgElement'

export interface AnimationOption extends RendererOption {
  ms: number
}
export type FrameAnimation = (origin: Path[], loopIndex?: number) => Path[]

export class SvgAnimation extends Renderer {
  public ms: number
  private _stopId: number
  private _stop: (() => void) | null
  private _anim: FrameAnimation | null
  private _restorePaths: Path[]
  private _framesNumber: number | undefined

  /**d
   * Releation animate element
   * TODO: add easing option
   */
  private _repeatCount: string
  constructor(
    el: HTMLElement,
    { background, ms }: AnimationOption = { ms: 60 }
  ) {
    super(el, { background })
    this.ms = ms
    this._stop = null
    this._anim = null
    this._restorePaths = []
    this._stopId = 0
    this._repeatCount = 'indefinite'
  }

  /**
   * @param {FramaAnimation} fn
   * @param {{ frame?: number; repeat?: number }} opts
   * `frame` is the number of frames to animate
   * `repeat` is related for rapeatCount of animate element attribute.
   */
  public setAnimation(
    fn: FrameAnimation,
    {
      frames,
      repeatCount,
      ms,
    }: { frames?: number; repeatCount?: number | string; ms?: number } = {}
  ): void {
    this._anim = fn
    this._framesNumber = frames
    this._repeatCount = repeatCount ? `${repeatCount}` : 'indefinite'
    if (ms) this.ms = ms
  }

  public stop(): boolean {
    if (this._stop) {
      this._stop()
      return true
    }
    return false
  }

  public restore(): void {
    this.paths = this._restorePaths
    this.update()
  }

  public generateFrame(index?: number): Path[] {
    if (!this._anim) return this.paths
    return this._anim(
      this._restorePaths.map((p) => p.clone()),
      index
    )
  }

  public start(): void {
    // If do not this first, this cannot get the number of frames well.
    this.stop()
    this._registerRestorePaths()

    let index = 0
    let start: number | undefined
    const ms = this.ms
    const loopCount: number = this._getFramesNumber()
    const frame: FrameRequestCallback = (timestamp) => {
      if (ms !== this.ms) {
        this.restore()
        this.start()
        return
      }
      if (!start || timestamp - start > ms) {
        start = timestamp
        this.paths = this.generateFrame(index)
        this.update()
        index = index > loopCount ? 0 : index + 1
      }
      this._stopId = requestAnimationFrame(frame)
    }
    this._stopId = requestAnimationFrame(frame)
    this._stop = () => {
      cancelAnimationFrame(this._stopId)
      this._stop = null
    }
  }

  public toAnimationElement(): SVGSVGElement {
    // If the animation is stopped, read the currently displayed Svg data.
    // If stopped in the middle, SVG in that state is displayed
    if (!this._stop) {
      this._registerRestorePaths()
    }

    const loopNumber = this._getFramesNumber()
    const animPathsList: Path[][] = Array(loopNumber)
      .fill(null)
      .map((_: any, i: number) => this.generateFrame(i))

    const dur = loopNumber * (this.ms > 0 ? this.ms : 1) + 'ms'
    const t = 1 / loopNumber
    const keyTimes = `0;${Array(loopNumber)
      .fill(undefined)
      .map((_, i) => roundUp((i + 1) * t, 4) + '')
      .join(';')}`

    const createAnimationElement = (
      origin: Path,
      attributeName: string,
      defaultValue: string,
      getValue: ({
        origin,
        path,
      }: {
        origin: Path
        path?: Path
      }) => string | undefined
    ): SVGElement | null => {
      const animValues = animPathsList.map((ap) => {
        const path = ap.find((p) => p.attrs.id === origin.attrs.id)
        return getValue({ origin, path }) || defaultValue
      })
      // return null if value is same all.
      if (animValues.every((v) => v === defaultValue)) return null

      return createSvgChildElement('animate', {
        dur,
        keyTimes,
        attributeName,
        repeatCount: this._repeatCount,
        values: [defaultValue, ...animValues].join(';'),
      })
    }
    const animEls = this._restorePaths.map((p) => {
      const pEl = p.toElement()
      const dAnimEl = createAnimationElement(
        p,
        'd',
        p.getCommandString(),
        ({ origin, path }) =>
          path && path.commands.length > 0
            ? path.getCommandString()
            : origin.commands[0].toString()
      )
      if (dAnimEl) pEl.appendChild(dAnimEl)

      Object.entries({
        fill: 'fill',
        stroke: 'stroke',
        strokeWidth: 'stroke-width',
      }).map(([propertyName, attrName]) => {
        const aEl = createAnimationElement(
          p,
          attrName,
          p[propertyName],
          ({ path }) => (path ? path[propertyName] : undefined)
        )
        if (aEl) pEl.appendChild(aEl)
      })

      // TODO: Validate attrs
      // exclude id
      const { id, ...attrs } = p.attrs
      Object.keys(attrs).map((propertyName: string) => {
        const attrName = camel2kebab(propertyName)
        const aEl = createAnimationElement(
          p,
          attrName,
          p.attrs[propertyName],
          ({ path }) => path?.attrs[propertyName]
        )
        if (aEl) pEl.appendChild(aEl)
      })

      return pEl
    })

    return createSvgElement(
      { width: String(this.width), height: String(this.height) },
      animEls
    )
  }

  /**
   * @param filename
   * TODO: Support gif and apng
   */
  public downloadAnimation(filename?: string): void {
    download({
      data: svg2base64(this.toAnimationElement().outerHTML),
      extension: 'svg',
      filename,
    })
  }

  /**
   * @return {number}
   * Default value is total of commands length.
   */
  private _getFramesNumber(): number {
    return this._framesNumber && this._framesNumber > 0
      ? this._framesNumber
      : this._restorePaths.reduce((l, p) => l + p.commands.length, 0)
  }

  private _registerRestorePaths() {
    this._restorePaths = this.clonePaths().map((p, i) => {
      p.attrs.id = `t${i}`
      return p
    })
  }
}
