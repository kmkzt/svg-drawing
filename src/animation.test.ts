import { SvgAnimation, FrameAnimation } from './animation'

const defaultTestData = `<svg width="200" height="200">
  <path fill="#f00" stroke="#00f" stroke-width="4" d="M 1 1 L 2 2 C 3 3 5 3 7 3 Z"></path>
  <path fill="#ff0" stroke="#f0f" stroke-width="2" d="M 2 2 L 4 4 C 6 6 10 6 14 6 Z"></path>
</svg>`

describe('animation.ts', () => {
  describe('SvgAnimation', () => {
    const generateAniamtion = (svgStr = defaultTestData) => {
      const svg = new SvgAnimation(document.createElement('div'))
      svg.parseSVGString(svgStr)
      return svg
    }
    it('init', () => {
      expect(generateAniamtion()).toMatchSnapshot()
    })
    it('generateFrame', () => {
      const svg = generateAniamtion()
      svg.setAnimation(paths => {
        return [paths[0]]
      })
      expect(svg.generateFrame().length).toBe(1)
    })
    // TODO: Improve test pattern
    it('toAnimationElement', () => {
      const svg = generateAniamtion()
      const testAnimation: FrameAnimation = (paths, count) => {
        const update = []
        for (let i = 0; i < paths.length; i += 1) {
          // Test property
          if (count % 2 === 0) paths[i].stroke = '#0ff'
          // Test Attribute
          if (count % 3 === 0)
            Object.assign(paths[i].attrs, {
              strokeLinecap: 'mitter'
            })
          if (count < paths[i].commands.length) {
            paths[i].commands = paths[i].commands.slice(0, count)
            update.push(paths[i])
            break
          }
          count -= paths[i].commands.length
          update.push(paths[i])
        }
        return update
      }
      svg.setAnimation(testAnimation)
      expect(svg.toAnimationElement()).toMatchSnapshot()
    })
  })
})
