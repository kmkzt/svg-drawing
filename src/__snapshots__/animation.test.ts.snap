// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`animation.ts SvgAnimation init 1`] = `
SvgAnimation {
  "_anim": null,
  "_paths": Array [
    Path {
      "attrs": Object {
        "strokeLinecap": "round",
        "strokeLinejoin": "round",
      },
      "commands": Array [
        Command {
          "type": "M",
          "value": Array [
            1,
            1,
          ],
        },
        Command {
          "type": "L",
          "value": Array [
            2,
            2,
          ],
        },
        Command {
          "type": "C",
          "value": Array [
            3,
            3,
            5,
            3,
            7,
            3,
          ],
        },
        Command {
          "type": "Z",
          "value": Array [],
        },
      ],
      "fill": "#f00",
      "stroke": "#00f",
      "strokeWidth": 4,
    },
    Path {
      "attrs": Object {
        "strokeLinecap": "round",
        "strokeLinejoin": "round",
      },
      "commands": Array [
        Command {
          "type": "M",
          "value": Array [
            2,
            2,
          ],
        },
        Command {
          "type": "L",
          "value": Array [
            4,
            4,
          ],
        },
        Command {
          "type": "C",
          "value": Array [
            6,
            6,
            10,
            6,
            14,
            6,
          ],
        },
        Command {
          "type": "Z",
          "value": Array [],
        },
      ],
      "fill": "#ff0",
      "stroke": "#f0f",
      "strokeWidth": 2,
    },
  ],
  "_repeatCount": "indefinite",
  "_restorePaths": Array [],
  "_stop": null,
  "_stopId": 0,
  "background": "#fff",
  "el": <div>
    <svg
      height="0"
      version="1.1"
      width="0"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    />
  </div>,
  "height": 0,
  "left": 0,
  "ms": 60,
  "top": 0,
  "width": 0,
}
`;

exports[`animation.ts SvgAnimation toAnimationElement 1`] = `
<svg
  height="0"
  version="1.1"
  width="0"
  xmlns="http://www.w3.org/2000/svg"
  xmlns:xlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M 1 1 L 2 2 C 3 3 5 3 7 3 Z"
    fill="#f00"
    id="t0"
    stroke="#00f"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="4"
  >
    <animate
      attributeName="d"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="M 1 1 L 2 2 C 3 3 5 3 7 3 Z;M 1 1;M 1 1;M 1 1 L 2 2;M 1 1 L 2 2 C 3 3 5 3 7 3;M 1 1 L 2 2 C 3 3 5 3 7 3 Z;M 1 1 L 2 2 C 3 3 5 3 7 3 Z;M 1 1 L 2 2 C 3 3 5 3 7 3 Z;M 1 1 L 2 2 C 3 3 5 3 7 3 Z"
    />
    <animate
      attributeName="stroke"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="#00f;#0ff;#00f;#0ff;#00f;#0ff;#00f;#0ff;#00f"
    />
    <animate
      attributeName="stroke-linecap"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="round;mitter;round;round;mitter;round;round;mitter;round"
    />
  </path>
  <path
    d="M 2 2 L 4 4 C 6 6 10 6 14 6 Z"
    fill="#ff0"
    id="t1"
    stroke="#f0f"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
  >
    <animate
      attributeName="d"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="M 2 2 L 4 4 C 6 6 10 6 14 6 Z;M 2 2;M 2 2;M 2 2;M 2 2;M 2 2;M 2 2;M 2 2 L 4 4;M 2 2 L 4 4 C 6 6 10 6 14 6"
    />
    <animate
      attributeName="stroke"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="#f0f;#f0f;#f0f;#f0f;#f0f;#0ff;#f0f;#0ff;#f0f"
    />
    <animate
      attributeName="stroke-linecap"
      dur="480ms"
      keyTimes="0;0.125;0.25;0.375;0.5;0.625;0.75;0.875;1"
      repeatCount="indefinite"
      values="round;round;round;round;round;mitter;round;round;mitter"
    />
  </path>
</svg>
`;
